README.md

```diff
@@ -1,8 +1,246 @@
+
+```dart
+// tool/dev.dart
+import 'package:dart_dev/dart_dev.dart';
+import 'package:glob/glob.dart';
+
+final config = {
+  'analyze': AnalyzeTool()
+    ..dartanalyzerArgs = ['--fatal-infos'],
+  'format': FormatTool()
+    ..exclude = [Glob('lib/src/generated/**.dart')],
+  'test': TestTool()
+    ..testArgs = ['-P', 'unit'],
+  'serve': WebdevServeTool()
+    ..webdevArgs = ['--debug', 'example:8080', 'web:8081'],
+};
+```
+
+## Motivation & Goal
+
+Most Dart projects eventually share a common set of development requirements
+(e.g. static analysis, formatting, test running, serving, etc.). The Dart SDK
+along with some core packages supply the necessary tooling for these developer
+tasks (e.g. `dartanalyzer`, `dartfmt`, or `pub run test`).
+
+While the core tooling gets us far, there are two areas in which we feel it
+falls short:
+
+1. Inconsistencies across projects in how these tools must be used in order to
+   to accomplish common developer tasks.
+
+2. Functionality gaps for more complex use cases.
+
+With `dart_dev`, we attempt to address #1 by providing a way to configure all of
+these common developer tasks at the project level, and #2 by composing
+additional functionality around existing tools.
+
+This package is built with configurability and extensiblity in mind, with the
+hope that you and your teams will find value in creating your own tools and
+shared configurations. Ideally, you or your team can settle on a shared
+configuration that individual projects can consume; projects with unique
+requirements can tweak the configuration as necessary; and developers can rely
+on the convention of a simple, consistent command-line interface regardless of
+the project they are in.
+
+## Project-Level Configuration
+
+Every task should be able to be configured at the project-level so that any
+variance across projects becomes a configuration detail that need not be
+memorized or referenced in order to run said task.
+
+Consider formatting as an example. The default approach to formatting files is
+to run `dartfmt -w .`. But, some projects may want to exclude certain files that
+would otherwise be formatted by this command. Or, some projects may want to use
+`pub run dart_style:format` instead of `dartfmt`. Currently, there is no
+project-level configuration supported by the formatter, so these sorts of things
+just have to be documented in a `README.md` or `CONTRIBUTING.md`.
+
+With `dart_dev`, this can be accomplished like so:
+
+```dart
+// tool/dev.dart
+import 'package:dart_dev/dart_dev.dart';
+
+final config = {
+  'format': FormatTool()
+    ..exclude = [Glob('lib/src/**.g.dart')]
+    ..formatter = Formatter.dartStyle,
+};
+```
+
+```bash
+$ ddev format
+[INFO] Running subprocess:
+pub run dart_style:format -w <3 paths>
+--------------------------------------
+Unchanged ./lib/foo.dart
+Unchanged ./lib/src/bar.dart
+Formatted ./lib/src/baz.dart
+```
+
+## Extending/Composing Functionality
```

Comment:
Are there any other commands that will change the underlying process run based on direct deps? If so, I think we should document them all here, since it's not very intuitive based on how everything else works. First time I ran this I was definitely trying to figure out what `TestTool()` option I had to add to use `build_runner` instead of `pub run test`.

Reply:
Yeah I do want to include more documentation around the individual tools, but I was hoping to keep that info out of the readme and in the api docs for each tool. In general I was just trying to keep the readme focused on the overall idea behind dart_dev (the current readme is pretty huge and tries to document everything but it is kind of a mess because of that).

Mind if I cut a follow up ticket to expand the docs on each of the tools and then include a more prominent reference to those docs in the readme?

Reply:
Sounds good